### PR TITLE
cpan and cran resources: check perl and R module installation

### PR DESCRIPTION
**Pull request by mgrobelin**
This resource allows to check Perl modules installed by system packages or CPAN.

```
control 'perl-modules-01' do
  impact 1.0
  title 'Perl modules available'
  desc 'Various Perl modules should be available.'
  
  describe cpan('DBI') do
    it { should be_installed }
  end
  
  describe cpan('DBD::Pg') do
    it { should be_installed }
    its('version') { should cmp >= '3.7.0' }
  end
  
  describe cpan('YAML::XS', '/home/jdoe/perl5/lib/perl5') do
    it { should be_installed }
  end
  
end
```

Perl module versions are detected by printing the individual modules VERSION object.

I thought to name this resource perl_module instead, but cpan aligns better with the other package related resource like gem, pip, npm etc.

### CODE REVIEW

File: docs/resources/cpan.md.erb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+<%= partial "/shared/matcher_be" %>
+
+### be_installed
+
+The `be_installed` matcher tests if the named package is installed on the system:
+
+    it { should be_installed }
+
+### cmp
```
Comment (adamleff) [inline docs/resources/cpan.md.erb:37]: You can remove the boilerplate matchers - we have since removed them from all other resource documentation since you submitted this.  Lines 27-30 and lines 37-52 can be deleted!

File: docs/resources/cran.md.erb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+* `'package_name'` is the name of the package, such as `'DBI'`
+* `be_installed` tests to see if the package described above is installed
+
+
+## Matchers
+
+This InSpec audit resource has the following matchers:
+
+### be
```
Comment (adamleff) [inline docs/resources/cran.md.erb:27]: Same as with the `cpan.md.erb` file - the boilerplate matcher documentation can be removed from this as well.

File: lib/resources/cpan.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      end
+    "
+
+    def initialize(package_name, perl_lib_path = nil)
+      @package_name = package_name
+      @perl_lib_path = perl_lib_path
+      @perl_cmd = 'perl'
+
+      # this resource is not supported on Windows
+      skip_resource 'The `cpan` resource is not supported on your OS yet.' if inspec.os.windows?
```
Comment (adamleff) [inline lib/resources/cpan.rb:27]: Can you add a `return` to the beginning of this so we don't run an unnecessary command existence check for perl on Windows systems?

File: lib/resources/cpan.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    def info
+      return @info if defined?(@info)
+
+      @info = {}
+      @info[:type] = 'cpan'
+      # set PERL5LIB environment variable if a custom lib path is given
+      lib_path = @perl_lib_path.nil? ? '' : "PERL5LIB=#{@perl_lib_path} "
+      cmd = inspec.command("#{lib_path+@perl_cmd} -le 'eval \"require $ARGV[0]\" and print $ARGV[0]->VERSION or exit 1' #{@package_name}")
+      return @info if cmd.exit_status != 0
```
Comment (adamleff) [inline lib/resources/cpan.rb:39]: Another alternative which we've been using more in our resources:

```ruby
return @info unless cmd.exit_status.zero?
```

File: lib/resources/cpan.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+      @info[:name] = @package_name
```
Comment (adamleff) [inline lib/resources/cpan.rb:41]: Perhaps we should move this up around line 35 so, if the command fails, we still return an `@info` hash that has more than just the `:type` key? Not a big deal, but nice for clarity if someone is debugging.

File: lib/resources/cran.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  class CranPackage < Inspec.resource(1)
+    name 'cran'
+    desc 'Use the `cran` InSpec audit resource to test R modules that are installed from CRAN package repository.'
+    example "
+      describe cran('DBI') do
+        it { should be_installed }
+      end
+    "
+
+    def initialize(package_name, perl_lib_path = nil)
```
Comment (adamleff) [inline lib/resources/cran.rb:21]: We can remove the `perl_lib_path` method parameter here, right?

File: lib/resources/cran.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      @package_name = package_name
+      @r_cmd = 'Rscript'
+
+      # this resource is not supported on Windows
+      skip_resource 'The `cran` resource is not supported on your OS yet.' if inspec.os.windows?
```
Comment (adamleff) [inline lib/resources/cran.rb:26]: Same as above - can we add a `return` here so we don't check for the command's existence on windows?

File: lib/resources/cran.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      return skip_resource 'Rscript not found' unless inspec.command(@r_cmd).exist?
+    end
+
+    def info
+      return @info if defined?(@info)
+
+      @info = {}
+      cmd = inspec.command("#{@r_cmd} -e 'packageVersion(\"#{@package_name}\")'")
+      return @info if cmd.exit_status != 0
```
Comment (adamleff) [inline lib/resources/cran.rb:35]: Same as above re `unless ... zero?`

File: lib/resources/cran.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+      # ... but just capture the first match for the moment
+      params = /^\[\d+\]\s+(?:\p{Initial_Punctuation})(.+)(?:\p{Final_Punctuation})$/.match(cmd.stdout.chomp)
+      @info = {
+          installed: !params.nil?,
+          type: 'cran',
+      }
+      return @info unless @info[:installed]
+
+      @info[:name] = @package_name
```
Comment (adamleff) [inline lib/resources/cran.rb:56]: Same as above... populate the name in the hash even if we can't find the version, just for clarity/transparency.

File: test/unit/mock/cmd/r-print-version [hunk 10]
```diff
@@ -0,0 +1 @@
+[1] ‘0.5.1’
```
Comment (adamleff) [inline test/unit/mock/cmd/r-print-version:1]: Since you mention it in your code comments, can we add multiple lines here to match the edge case you refer to in order to ensure we're pulling in the version string you expect?

File: test/unit/resources/cpan_test.rb [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+# author: Christoph Hartmann
+# author: Dominik Richter
+
+require 'helper'
+require 'inspec/resource'
+
+describe 'Inspec::Resources::Cpan' do
+  it 'verify cpan package detail parsing' do
+    resource = load_resource('cpan', 'DBD::Pg')
+      pkg = {:type=>'cpan', :name=>'DBD::Pg', :version=>'3.7.0', :installed=>true }
```
Comment (adamleff) [inline test/unit/resources/cpan_test.rb:11]: Looks like indentation is off here. Also, you can write this with the newer non-hash-rocket hash syntax:

```ruby
pkg = {type: 'cpan', name: 'DBD::Pg', version: '3.7.0', installed: true}
```

File: test/unit/resources/cpan_test.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# encoding: utf-8
+# author: Christoph Hartmann
+# author: Dominik Richter
+
+require 'helper'
+require 'inspec/resource'
+
+describe 'Inspec::Resources::Cpan' do
+  it 'verify cpan package detail parsing' do
```
Comment (adamleff) [inline test/unit/resources/cpan_test.rb:9]: We should add a negative test as well, ensuring it returns the appropriate messages when a package is not installed. Let me know if you need guidance on doing that.

**Review by adamleff — CHANGES_REQUESTED** [review]
Hi, @mgrobelin! Thank you for your first contribution to InSpec! Welcome to the community!

I've left you some feedback on your PR that we'll need to address before we can move forward. In addition, it appears that there are no unit tests for the `cran` resource, so we'll need to add some coverage there too.

Your change is currently failing some tests - see the Travis or AppVeyor output for more information. Essentially, lint checking is failing due to some Rubocop errors.

It was a bit of a surprise to see the PR title indicating this was a new resource for `cpan` but also seeing `cran` in it as well. While I love seeing both of these additions to InSpec, it's best to keep the PRs separate to ensure they get proper review.

One last thing: our project adheres to the Developer Certificate of Origin which requires each of your commits to include a [DCO sign-off](https://github.com/chef/inspec/blob/master/CONTRIBUTING.md#dco-sign-off-methods). It would be fine, while you are addressing the PR review feedback, to rebase and squash your commits down to a single commit and amend the message to include the sign-off. Let me know if you have any questions on that.

Again, thank you for your contribution! I'm looking forward to helping you push this across the finish line.

**Comment by mgrobelin** [comment]
Thanks for feedback @adamleff 

I will try to create two new PR's without the mess

**Comment by adamleff** [comment]
No worries @mgrobelin - we can reuse this PR if it's easier as well... I was just asking for future contributions to help keep the PRs separate :)

If I can be of any assistance, please let me know.
